### ISSUE TITLE
Embedding changes even though gradient is zero when using Adam

### ISSUE THREAD

**Issue by jgontrum** [pytorch/pytorch#18404]
## 🐛 Bug

I discovered that the embeddings of words slightly change after calling `optimizer.step()`, even though the embedding was not part of the back propagation chain and its gradients were zero. Interestingly, this only happens after the word was seen once, but occurs at every update afterward. If I use SGD instead of Adam as an optimizer, my code behaves as expected.

To me this smells a lot like a bug, but if this is intended behaviour, could you please explain why it is happing and what the motivation for it is?

**I call `zero_grad()` at the beginning of every batch, both for the optimizer and the model itself!**
## To Reproduce

1. Create a task that uses embeddings.
1. Use Adam optimizer.
1. Pick an infrequent word that is not part of every batch.
1. Check the embedding weigths and gradients before and after an update of the network.
1. After the word was updated once, its weights will continue to change after every batch, no matter if it was part of it or not.

You can find a naive minimal example as a Gist [here](https://gist.github.com/jgontrum/46b14b27ea20dc238358d104d793262f) or as a Google Colab notebook [here](https://colab.research.google.com/drive/1MeaqdY8bELJYpZBvfEov7qKlCExJ5CkA).

In my example I have four sentences, and the example word (‘better’ in this case) only occurs in sentence number three. In the first epoch, the embedding of for ‘better’ does not change for sentences one and two (as expected). In sentence three it is updated, also as expected. However, from now on it continuously changes at every update, no matter for what sentence:

This is a part of the output of my minimal example:
```
----------------------------
Starting epoch #1
Current sentence: 'this is a very good sentence number one' 
✅ The sentence did not contain the word 'better' and its embedding was not updated. Hurray.

Current sentence: 'a rather bad example' 
✅ The sentence did not contain the word 'better' and its embedding was not updated. Hurray.

Current sentence: 'an even better sentence to use' 
✅ The sentence contained the word 'better' and its embedding was updated. Hurray.

Current sentence: 'negative sentence very short' 
❌ The sentence did not contain the word 'better', but the embedding was updated anyhow. Boo.
The gradient for the embedding:  [0. 0. 0. 0. 0.]
The embedding before the update: [-1.3233105   0.5050922  -0.12209229 -0.27128008  0.9851864 ]
The embedding after the update:  [-1.3238337   0.50561535 -0.12261546 -0.27075687  0.9846632 ]
Summed difference between both embeddings: 0.0005232319235801697

----------------------------
Starting epoch #2
Current sentence: 'this is a very good sentence number one' 
❌ The sentence did not contain the word 'better', but the embedding was updated anyhow. Boo.
The gradient for the embedding:  [0. 0. 0. 0. 0.]
The embedding before the update: [-1.3238337   0.50561535 -0.12261546 -0.27075687  0.9846632 ]
The embedding after the update:  [-1.3242759   0.50605756 -0.12305766 -0.27031463  0.984221  ]
Summed difference between both embeddings: 0.00044211000204086304

Current sentence: 'a rather bad example' 
❌ The sentence did not contain the word 'better', but the embedding was updated anyhow. Boo.
The gradient for the embedding:  [0. 0. 0. 0. 0.]
The embedding before the update: [-1.3242759   0.50605756 -0.12305766 -0.27031463  0.984221  ]
The embedding after the update:  [-1.324657    0.5064387  -0.12343878 -0.2699335   0.98383987]
Summed difference between both embeddings: 0.0003810897469520569

Current sentence: 'an even better sentence to use' 
✅ The sentence contained the word 'better' and its embedding was updated. Hurray.

Current sentence: 'negative sentence very short' 
❌ The sentence did not contain the word 'better', but the embedding was updated anyhow. Boo.
The gradient for the embedding:  [0. 0. 0. 0. 0.]
The embedding before the update: [-1.3252577   0.5070339  -0.12404114 -0.2693513   0.9832365 ]
The embedding after the update:  [-1.3257872   0.5075586  -0.12457216 -0.26883805  0.9827046 ]
Summed difference between both embeddings: 0.000554502010345459
```

## Expected behavior

I expect the embedding of an entry to only change after it was part of a back propagation chain and stay the same otherwise.

## Environment

PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.6
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.2
[pip3] numpydoc==0.8.0
[pip3] torch==1.0.1.post2
[pip3] torchvision==0.2.2.post3
[pip3] torchviz==0.0.1
[conda] Could not collect

## Additional context

You can find a naive minimal example as a Gist [here](https://gist.github.com/jgontrum/46b14b27ea20dc238358d104d793262f) or as a Google Colab notebook [here](https://colab.research.google.com/drive/1MeaqdY8bELJYpZBvfEov7qKlCExJ5CkA).

**Comment by honnibal** [comment]
Whether you consider this a bug or not depends on somewhat subjective interpretation of what the layer should be doing. The adam solver uses a moving average of the gradients, so even if you have a 0 gradient in the batch you can have a non-zero update. If PyTorch supports sparse updates for embedding layers, that would probably solve the issue.

I discussed this a while ago on the DyNet tracker: https://github.com/clab/dynet/issues/70 . DyNet does support choice of sparse or dense updates, so you can compare results there if you like. I don't think it makes a big difference.

**Comment by Evpok** [comment]
> If PyTorch supports sparse updates for embedding layers, that would probably solve the issue.
It does actually (setting `sparse=True` when building the `Embedding` layer), but it requires a separate optimizer for these layers (`SparseAdam`) or AllenNLP `DenseSparseAdam` (which is a wrapper for both versions of Adam)

**Comment by soumith** [comment]
As mentioned by the folks above, we indeed provide SparseAdam that explicitly defines how we modify the Adam algorithm to deal with sparse situations.

As @honnibal mentioned, this is not really a bug if you actually use the original formulation as-is.

See https://pytorch.org/docs/stable/optim.html#torch.optim.SparseAdam for more info
